data/transactions/logic/eval.go

```diff
@@ -2571,6 +2571,81 @@ func opSetByte(cx *evalContext) {
 	cx.stack = cx.stack[:prev]
 }
 
+func extract(x []byte, start, length int) (out []byte, err error) {
+	out = x
+	end := start + length
+	if start > len(x) || end > len(x) {
+		err = errors.New("extract range beyond length of string")
+		return
+	}
+	out = x[start:end]
+	err = nil
+	return
+}
+
+func opExtract(cx *evalContext) {
+	last := len(cx.stack) - 1
+	start := cx.program[cx.pc+1]
+	length := cx.program[cx.pc+2]
+	cx.stack[last].Bytes, cx.err = extract(cx.stack[last].Bytes, int(start), int(length))
+}
+
+func opExtract3(cx *evalContext) {
+	last := len(cx.stack) - 1 // length
+	prev := last - 1          // start
+	pprev := prev - 1         // bytes
+	start := cx.stack[prev].Uint
+	length := cx.stack[last].Uint
+	if start > math.MaxInt32 || length > math.MaxInt32 {
+		cx.err = errors.New("extract range beyond length of string")
+		return
+	}
+	cx.stack[pprev].Bytes, cx.err = extract(cx.stack[pprev].Bytes, int(start), int(length))
+	cx.stack = cx.stack[:prev]
+}
+
+func convertBytesToInt(x []byte) (out uint64) {
+	out = uint64(0)
+	for _, b := range x {
+		out = out << 8
+		out = out | (uint64(b) & 0x0ff)
+	}
+	return
+}
+
+func opExtract16Bits(cx *evalContext) {
+	last := len(cx.stack) - 1 // start
+	prev := last - 1          // bytes
+	start := cx.stack[last].Uint
+	cx.stack[prev].Bytes, cx.err = extract(cx.stack[prev].Bytes, int(start), 2) // extract 16 bits
+
+	cx.stack[prev].Uint = convertBytesToInt(cx.stack[prev].Bytes)
```

Comment:
It looks like you are doing the conversion, so my earlier comment should instead be a note to change the spec to say so.